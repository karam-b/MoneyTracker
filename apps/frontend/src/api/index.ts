/* eslint-disable no-empty-pattern */
// todo: once openapi is implemented, this file could auto-generated
import { Action } from '@src/lib/react-query'
import {
  SchemaCategoryIn,
  SchemaCategoryOut,
  SchemaLogIn,
  SchemaLogOut,
  SchemaProfile,
} from 'types/dist/ts/schema'
import {
  RoutesAuthLocalLogin,
  RoutesAuthLocalRegister,
  RoutesEmailStatus,
  RoutesProfileUpdate,
  RoutesUpdatePasswordLocal,
  RoutesUpdatePasswordNolocal,
} from 'types/dist/ts/routes'

export const find_log: Action<object, SchemaLogOut[]> =
  //
  () => ({
    path: '/log',
    method: 'GET',
  })

export const find_one_log: Action<{ _id: string }, SchemaLogOut> =
  //
  ({ _id }: { _id: string }) => ({
    path: '/log/' + _id,
    method: 'GET',
  })

export const create_log: Action<SchemaLogIn, SchemaLogOut> =
  //
  data => ({
    path: '/log',
    method: 'POST',
    body: JSON.stringify(data),
  })

export const update_log: Action<
  Partial<SchemaLogIn> & { _id: string },
  SchemaLogOut
> =
  //
  ({ _id, ...data }) => ({
    path: '/log/' + _id,
    method: 'PUT',
    body: JSON.stringify(data),
  })

export const delete_log: Action<{ _id: string }, null> =
  //
  ({ _id }) => ({
    path: '/log/' + _id,
    method: 'DELETE',
  })

export const find_category: Action<object, SchemaCategoryOut[]> =
  //
  () => ({
    path: '/category',
    method: 'GET',
  })

export const find_one_category: Action<{ _id: string }, SchemaCategoryOut> =
  //
  ({ _id }) => ({
    path: '/category/' + _id,
    method: 'GET',
  })

export const create_category: Action<SchemaCategoryIn, SchemaCategoryOut> =
  //
  data => ({
    path: '/category',
    method: 'POST',
    body: JSON.stringify(data),
  })

export const update_category: Action<
  Partial<SchemaCategoryIn> & { _id: string },
  SchemaCategoryOut
> =
  //
  ({ _id, ...data }) => ({
    path: '/category/' + _id,
    method: 'PUT',
    body: JSON.stringify(data),
  })

export const delete_category: Action<{ _id: string }, null> =
  //
  ({ _id }) => ({
    path: '/category/' + _id,
    method: 'DELETE',
  })

// todo: should be removed in favor of filtering feature
export const find_all_logs_by_category: Action<
  { _id: string },
  SchemaLogOut[]
> =
  //
  ({ _id }) => ({
    path: '/category/' + _id + '/logs',
    method: 'GET',
  })

// todo: in formik I pass confirmPassword, here this will not get passed
// but in different actions that uses `JSON.stringify(data)` it will be passed
export const register: Action<
  RoutesAuthLocalRegister & RoutesAuthLocalLogin,
  SchemaProfile
> =
  //
  ({ displayName, ...data }) => ({
    path: '/auth/local/register',
    method: 'POST',
    body: JSON.stringify({ displayName }),
    headers: {
      Authorization: 'Basic ' + btoa(data.email + ':' + data.password),
    },
  })

export const login: Action<RoutesAuthLocalLogin, SchemaProfile> =
  //
  data => ({
    path: '/auth/local/login',
    method: 'POST',
    headers: {
      Authorization: 'Basic ' + btoa(data.email + ':' + data.password),
    },
  })

export const email_status: Action<RoutesEmailStatus, ('local' | 'google')[]> =
  //
  data => ({
    path: '/profile/status',
    method: 'GET',
    body: JSON.stringify(data),
  })

export const profile: Action<object, SchemaProfile> =
  //
  () => ({
    path: '/profile',
    method: 'GET',
  })

export const update_profile: Action<RoutesProfileUpdate, SchemaProfile> =
  //
  data => ({
    path: '/profile',
    method: 'PUT',
    body: JSON.stringify(data),
  })

export const set_password: Action<
  RoutesUpdatePasswordLocal | RoutesUpdatePasswordNolocal,
  null
> =
  //
  data => ({
    path: '/profile/password',
    method: 'PUT',
    body: JSON.stringify(data),
  })
