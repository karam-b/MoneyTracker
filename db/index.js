/* eslint-disable @typescript-eslint/no-var-requires */
// @ts-check
/* eslint-env node */
const assert = require("assert");
const express = require("express");
const { parse } = require("pg-connection-string");
const { Client } = require("pg");
const _ = require("express-async-handler");

if (process.env.NODE_ENV !== "test") {
  throw new Error("this script is only for testing environments");
}

assert(process.env.PORT, "no port provided");
assert(process.env.DATABASE_URL, "no string provided");
assert(process.env.POSTGRES_DB, "no db provided");

const config = parse(process.env.DATABASE_URL || "");
const client = new Client({
  user: config.user,
  database: process.env.POSTGRES_DB,
  password: config.password,
  port: Number(config.port || 5432),
});

async function main() {
  await client.connect();
  console.log("db connected");

  const app = express();

  app.get(
    "/clean_db",
    _(async (req, res) => {
      await client.query("DELETE FROM users;");

      await client.query("DELETE FROM entry;");

      await client.query("DELETE FROM category;");

      res.send("db cleaned");
    })
  );

  app.get(
    "/remove_user_data/:user_id",
    _(async (req, res) => {
      const user_id = Number(req.params.user_id);
      await client.query("DELETE FROM entry WHERE created_by = $1;", [user_id]);

      await client.query("DELETE FROM category WHERE created_by = $1;", [
        user_id,
      ]);

      await client.query("DELETE FROM users WHERE id = $1;", [user_id]);

      res.send("user data removed");
    })
  );

  app.use((err, req, res, next) => {
    console.error(err);
    res.status(500).send("Something broke!");
  });

  app.use((req, res, next) => {
    console.log("helper server: sorry cant find that", req.url);
    res.status(404).send("Sorry cant find that!");
  });

  app.listen(process.env.PORT, () => {
    console.log("helper server: listening on " + process.env.PORT);
  });
}

main().catch((err) => {
  console.error(err);
  process.exit(1);
});
