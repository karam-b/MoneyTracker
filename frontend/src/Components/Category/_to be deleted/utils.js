// default icon
import Icon_cat from "@material-ui/icons/Category";
// catagory icons
import Icon_Fastfood from "@material-ui/icons/Fastfood"; // fastFood
import Icon_Kitchen from "@material-ui/icons/Kitchen"; //localFood
import Icon_LocalCafe from "@material-ui/icons/LocalCafe"; //cafe
import Icon_LocalBar from "@material-ui/icons/LocalBar"; //bar
// transtspot
import Icon_LocalGasStation from "@material-ui/icons/LocalGasStation"; //gas
import Icon_DirectionsCar from "@material-ui/icons/DirectionsCar"; //personalCar
import Icon_Flight from "@material-ui/icons/Flight"; //flight
import Icon_Train from "@material-ui/icons/Train"; //publicTransportation
// shopping
import Icon_ShoppingCart from "@material-ui/icons/ShoppingCart"; //shopping
import Icon_ShoppingBasket from "@material-ui/icons/ShoppingBasket"; //groceries
// Entertainment
import Icon_SportsEsports from "@material-ui/icons/SportsEsports"; //gaming
import Icon_SportsBaseball from "@material-ui/icons/SportsBaseball"; //sport
import Icon_CardGiftcard from "@material-ui/icons/CardGiftcard"; //gift

export const allAvailableIconJSX = {
  Fastfood: {
    JSX: <Icon_Fastfood/>
  },
  Kitchen: {
    JSX: <Icon_Kitchen/>
  },
  LocalCafe: {
    JSX: <Icon_LocalCafe/>
  },
  LocalBar: {
    JSX: <Icon_LocalBar/>
  },
  LocalGasStation: {
    JSX: <Icon_LocalGasStation/>
  },
  DirectionsCar: {
    JSX: <Icon_DirectionsCar/>
  },
  Flight: {
    JSX: <Icon_Flight/>
  },
  Train: {
    JSX: <Icon_Train/>
  },
  ShoppingCart: {
    JSX: <Icon_ShoppingCart/>
  },
  ShoppingBasket: {
    JSX: <Icon_ShoppingBasket/>
  },
  SportsEsports: {
    JSX: <Icon_SportsEsports/>
  },
  SportsBaseball: {
    JSX: <Icon_SportsBaseball/>
  },
  CardGiftcard: {
    JSX: <Icon_CardGiftcard/>
  },
}


// the color property should only be from these folowing colors
export const colors = [
  "amber",
  "blue",
  "blueGrey",
  "brown",
  "common",
  "cyan",
  "deepOrange",
  "deepPurple",
  "green",
  "grey",
  "indigo",
  "lightBlue",
  "lightGreen",
  "lime",
  "orange",
  "pink",
  "purple",
  "red",
  "teal",
  "yellow",
];
