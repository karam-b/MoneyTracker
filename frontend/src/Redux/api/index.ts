import category_create from './category_create'
import category_delete from './category_delete'
import category_find from './category_find'
import category_findOne from './category_findOne'
import category_update from './category_update'
import log_create from './log_create'
import log_delete from './log_delete'
import log_find from './log_find'
import log_findOne from './log_findOne'
import log_update from './log_update'
import profile_fetch from './profile_fetch'
import profile_password from './profile_password'
import profile_update from './profile_update'
import user_login from './user_login'
import user_signup from './user_signup'
import user_status from './user_status'
