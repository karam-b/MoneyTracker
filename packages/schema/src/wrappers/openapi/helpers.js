// @ts-check

/**
 * @function
 * @template Option
 * @param {import('./_').RawComponent<Option>} input
 * @returns {unknown extends Option ?
 *      () => import('./_').Component :
 *      (options: Option) => import('./_').Component
 *    }
 */
export function RawFunctionTransformer(input) {
  return (option) => (data, actions) => input({ data, actions }, option)
}
