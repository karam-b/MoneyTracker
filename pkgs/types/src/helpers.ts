/**
 * helper function to override properties that have different values in different contexts
 * ex: log.category maybe string before population and of type Cat after population
 */
export type Override<A, B extends Partial<A>> = {
  [K in keyof A]: B[K] extends A[K] ? B[K] : A[K]
}

/**
 * exclude any property that may have be of null type
 * ex: ExcludeNull<{a: number, b: boolean | null}> === {a: number}
 */
export type ExcludeNull<I, T extends keyof I = keyof I> = Pick<
  I,
  T extends T ? (null extends I[T] ? never : T) : never
>
