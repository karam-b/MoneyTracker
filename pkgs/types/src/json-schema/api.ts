import { JSONSchema7 } from 'json-schema'

export const jwt_payload = {
  $schema: 'http://json-schema.org/draft-07/schema#',
  $id: '/jwt' as const,
  type: 'object',
  properties: {
    _id: { type: 'string' },
    email: { type: 'string' },
    exp: { type: 'number' },
    iat: { type: 'number' },
  },
  required: ['_id', 'email', 'exp', 'iat'],
  additionalProperties: false,
} satisfies JSONSchema7

export const api_google_callback_params = {
  $schema: 'http://json-schema.org/draft-07/schema#',
  $id: '/api_google_callback_params' as const,
  type: 'object',
  properties: {
    token: { type: 'string' },
    displayName: { type: 'string' },
    _id: { type: 'string' },
  },
  required: ['token', 'displayName', '_id'],
  additionalProperties: false,
} satisfies JSONSchema7
