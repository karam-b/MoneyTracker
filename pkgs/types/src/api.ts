import { JSONSchema7 } from 'json-schema'

export const jwt_payload = {
  $schema: 'http://json-schema.org/draft-07/schema#',
  $id: 'https://example.com/jwt',
  type: 'object',
  properties: {
    _id: { type: 'string' },
    email: { type: 'string' },
    exp: { type: 'number' },
    iat: { type: 'number' },
  },
  required: ['_id', 'email', 'exp', 'iat'],
  additionalProperties: false,
} satisfies JSONSchema7

export const api_google_callback_params = {
  $schema: 'http://json-schema.org/draft-07/schema#',
  $id: 'https://example.com/api_google_callback_params',
  type: 'object',
  properties: {
    token: { type: 'string' },
    displayName: { type: 'string' },
    _id: { type: 'string' },
  },
  required: ['token', 'displayName', '_id'],
  additionalProperties: false,
} satisfies JSONSchema7
