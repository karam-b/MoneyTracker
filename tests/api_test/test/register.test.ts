import { faker } from '@faker-js/faker'
import { it, expect } from '@jest/globals'
import api from './helpers/api'
import { getRandomUser } from './helpers/utils'
import { userMatcher } from './helpers/matcher'
import { clean_user } from './helpers/db'

it('invalid token: not provided', async () => {
  faker.seed(11111)
  const info = getRandomUser()

  const test = await api('/auth/local/register', {
    method: 'POST',
    json: {
      displayName: info.displayName,
    },
  })

  expect(test).toMatchSnapshot()
})

it('invalid token: not formatted', async () => {
  faker.seed(11112)
  const info = getRandomUser()

  const test = await api('/auth/local/register', {
    method: 'POST',
    headers: {
      Authorization:
        'Basic ' +
        Buffer.from(`${info.email}${info.password}`).toString('base64'),
    },
    json: {
      displayName: info.displayName,
    },
  })

  expect(test).toMatchSnapshot()
})

it('success', async () => {
  faker.seed(11113)
  const info = getRandomUser()

  const test = await api('/auth/local/register', {
    method: 'POST',
    headers: {
      Authorization:
        'Basic ' +
        Buffer.from(`${info.email}:${info.password}`).toString('base64'),
    },
    json: {
      displayName: info.displayName,
    },
  })

  expect(test).toMatchSnapshot(userMatcher)
  // todo json doesn't work
  // expect(test.res?.data?.displayName).toBe(info.displayName)
  await clean_user(test.res?.data?._id)
})

it('success: only required', async () => {
  faker.seed(11114)
  const info = getRandomUser()

  const test = await api('/auth/local/register', {
    method: 'POST',
    headers: {
      Authorization:
        'Basic ' +
        Buffer.from(`${info.email}:${info.password}`).toString('base64'),
    },
  })

  expect(test).toMatchSnapshot(userMatcher)
  await clean_user(test.res?.data?._id)
})

it('user exist', async () => {
  faker.seed(11115)
  const info = getRandomUser()

  const old = await api('/auth/local/register', {
    method: 'POST',
    headers: {
      Authorization:
        'Basic ' +
        Buffer.from(`${info.email}:${info.password}`).toString('base64'),
    },
  })

  const test = await api('/auth/local/register', {
    method: 'POST',
    headers: {
      Authorization:
        'Basic ' +
        Buffer.from(`${info.email}:${info.password}`).toString('base64'),
    },
  })

  expect(test).toMatchSnapshot()
  await clean_user(old.res?.data?._id)
})
