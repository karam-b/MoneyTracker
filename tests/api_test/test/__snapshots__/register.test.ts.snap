// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`invalid token: not formatted 1`] = `
Object {
  "req": Object {
    "headers": Object {
      "Authorization": "Basic QW5keV9CYXJ0ZWxsMTJAZ21haWwuY29taUJabUZGNWtvWElvTWhG",
    },
    "json": Object {
      "displayName": "Demarco",
    },
    "method": "POST",
  },
  "res": Object {
    "data": null,
    "error": Object {
      "details": null,
      "message": "email and password not provided",
      "name": "BadBasicToken",
      "status": 401,
    },
  },
  "url": "/auth/local/register",
}
`;

exports[`invalid token: not provided 1`] = `
Object {
  "req": Object {
    "json": Object {
      "displayName": "Guy",
    },
    "method": "POST",
  },
  "res": Object {
    "data": null,
    "error": Object {
      "details": null,
      "message": "email and password not provided",
      "name": "BadBasicToken",
      "status": 401,
    },
  },
  "url": "/auth/local/register",
}
`;

exports[`success 1`] = `
Object {
  "req": Object {
    "headers": Object {
      "Authorization": "Basic RGVsZmluYS5TY2htaWR0MkB5YWhvby5jb206MTVXbXBybUh0ajBkNjlu",
    },
    "json": Object {
      "displayName": "Major",
    },
    "method": "POST",
  },
  "res": Object {
    "data": Object {
      "__v": 0,
      "_id": Any<String>,
      "createdAt": Any<String>,
      "displayName": Any<String>,
      "email": "Delfina.Schmidt2@yahoo.com",
      "providers": Array [
        "local",
      ],
      "token": Any<String>,
      "updatedAt": Any<String>,
    },
  },
  "url": "/auth/local/register",
}
`;

exports[`success: only required 1`] = `
Object {
  "req": Object {
    "headers": Object {
      "Authorization": "Basic SXNtYWVsLlBvbGxpY2hAZ21haWwuY29tOndmeHBqa0VtUzE4dGZuVQ==",
    },
    "method": "POST",
  },
  "res": Object {
    "data": Object {
      "__v": 0,
      "_id": Any<String>,
      "createdAt": Any<String>,
      "displayName": Any<String>,
      "email": "Ismael.Pollich@gmail.com",
      "providers": Array [
        "local",
      ],
      "token": Any<String>,
      "updatedAt": Any<String>,
    },
  },
  "url": "/auth/local/register",
}
`;

exports[`user exist 1`] = `
Object {
  "req": Object {
    "headers": Object {
      "Authorization": "Basic RGVyaWNrNTdAeWFob28uY29tOkNtQXVzZHRkNkV5cEZ2VA==",
    },
    "method": "POST",
  },
  "res": Object {
    "data": null,
    "error": Object {
      "details": Object {
        "errors": Object {
          "email": "this email is used",
        },
      },
      "message": "User already exist.",
      "name": "UserAlreadyExist",
      "status": 409,
    },
  },
  "url": "/auth/local/register",
}
`;
